### PR TITLE
Ft next id generate

### PR DESCRIPTION
**Pull request by mubarak23**
Draft PR for issue #274

### CODE REVIEW

**Comment by mubarak23** [comment]
getting specific state value not fix this error 
 @Marchand-Nicolas 
 
```
error[E0277]: the trait bound `fn(axum::extract::State<Arc<AppState>>, Extension<std::string::String>, axum::Json<create_balance::CreateBalance>) -> impl futures::Future<Output = impl IntoResponse> {create_balance::handler}: Handler<_, _, _>` is not satisfied
   --> src/endpoints/admin/balance/create_balance.rs:30:14
    |
29  | #[route(post, "/admin/tasks/balance/create", auth_middleware)]
    | -------------------------------------------------------------- required by a bound introduced by this call
30  | pub async fn handler(
    |              ^^^^^^^ the trait `Handler<_, _, _>` is not implemented for fn item `fn(State<Arc<AppState>>, Extension<String>, Json<CreateBalance>) -> impl Future<Output = impl IntoResponse> {handler}`
    |
    = help: the following other types implement trait `Handler<T, S, B>`:
              <Layered<L, H, T, S, B, B2> as Handler<T, S, B2>>
              <MethodRouter<S, B> as Handler<(), S, B>>
note: required by a bound in `post`
   --> /Users/macbookpro/.cargo/registry/src/index.crates.io-6f17d22bba15001f/axum-0.6.20/src/routing/method_routing.rs:407:1
    |
407 | top_level_handler_fn!(post, POST);
    | ^^^^^^^^^^^^^^^^^^^^^^----^^^^^^^
    | |                     |
    | |                     required by a bound in this function
    | required by this bound in `post`
    = note: this error originates in the macro `top_level_handler_fn` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by Marchand-Nicolas** [comment]
@mubarak23 Maybe check how we do in the other handlers to manage async functions calls

**Comment by Marchand-Nicolas** [comment]
Hello @mubarak23 OD hack ends on Sunday (6). Everything okay ?

**Comment by Marchand-Nicolas** [comment]
Hello, just got enough time to look at your problem @mubarak23 .

You are using Mutex from the std lib, which is not really usable inside of async functions. The best solution is to use Mutex from the Tokio Sync library: `use tokio::sync::Mutex;`

You can then simply do:
```rs
    let state_last_id = state.last_task_id.lock().await;
```

File: src/utils.rs [hunk 1]
```diff
@@ -840,3 +842,35 @@ pub fn parse_string(input: &str, address: FieldElement) -> String {
...
+    if let Some(doc) = last_doc {
+        let db_last_id = doc.id;
+
+        next_id = std::cmp::max(db_last_id as i32, (last_task_id).try_into().unwrap()) + 1;
+    } else {
+        // next_id = *state_last_id;
+        next_id = (last_task_id as i32) + 1;
+    }
+
+    // last_task_id = next_id.try_into().unwrap();
```
Comment (Marchand-Nicolas) [inline src/utils.rs]: Remove this comment

File: src/utils.rs [hunk 2]
```diff
@@ -840,3 +842,35 @@ pub fn parse_string(input: &str, address: FieldElement) -> String {
...
+        .unwrap();
+
+    let mut next_id: i32 = 1;
+
+    if let Some(doc) = last_doc {
+        let db_last_id = doc.id;
+
+        next_id = std::cmp::max(db_last_id as i32, (last_task_id).try_into().unwrap()) + 1;
+    } else {
+        // next_id = *state_last_id;
```
Comment (Marchand-Nicolas) [inline src/utils.rs]: Remove this comment

File: src/utils.rs [hunk 3]
```diff
@@ -840,3 +842,35 @@ pub fn parse_string(input: &str, address: FieldElement) -> String {
 
     result
 }
+
+// QuestInsertDocument
+pub async fn get_next_task_id(
+    // last_doc: Option<QuestTaskDocument>,
```
Comment (Marchand-Nicolas) [inline src/utils.rs]: Remove this comment

File: src/utils.rs [hunk 4]
```diff
@@ -840,3 +842,35 @@ pub fn parse_string(input: &str, address: FieldElement) -> String {
...
+    task_collection: &Collection<QuestTaskDocument>,
+    last_task_id: i64,
+) -> i32 {
+    // let mut state_last_id =state.last_task_id.lock().unwrap();
```
Comment (Marchand-Nicolas) [inline src/utils.rs]: Remove this comment

File: src/models.rs [hunk 5]
```diff
@@ -8,8 +8,11 @@ use starknet::{
 
 use crate::endpoints::quests::uri::Attribute;
 use crate::{config::Config, logger::Logger};
+use std::net::SocketAddr;
```
Comment (Marchand-Nicolas) [inline src/models.rs]: No need to import this module

File: src/utils.rs [hunk 6]
```diff
@@ -31,6 +32,7 @@ use std::hash::{Hash, Hasher};
 use std::result::Result;
 use std::str::FromStr;
 use std::{fmt::Write, sync::Arc};
+use tokio::sync::Mutex;
```
Comment (Marchand-Nicolas) [inline src/utils.rs]: No need to import that

File: src/endpoints/admin/balance/create_balance.rs [hunk 7]
```diff
@@ -44,11 +46,15 @@ pub async fn handler(
...
-    let mut next_id = 1;
-    if let Some(doc) = last_doc {
-        let last_id = doc.id;
-        next_id = last_id + 1;
-    }
+    // let mut next_id = 1;
+    // if let Some(doc) = last_doc {
+    //     let last_id = doc.id;
+    //     next_id = last_id + 1;
+    // }
```
Comment (Marchand-Nicolas) [inline src/endpoints/admin/balance/create_balance.rs]: No need to keep these comments

File: src/endpoints/admin/balance/create_balance.rs [hunk 8]
```diff
@@ -35,6 +36,7 @@ pub async fn handler(
     // Get the last id in increasing order
     let last_id_filter = doc! {};
     let options = FindOneOptions::builder().sort(doc! {"id": -1}).build();
+
     let last_doc = &collection.find_one(last_id_filter, options).await.unwrap();
```
Comment (Marchand-Nicolas) [inline src/endpoints/admin/balance/create_balance.rs]: You can remove this variable

Comment (Marchand-Nicolas) [reply src/endpoints/admin/balance/create_balance.rs]: Do the same in each route you modifed

File: src/endpoints/admin/quiz/create_question.rs [hunk 9]
```diff
@@ -1,9 +1,10 @@
+use crate::middleware::auth::auth_middleware;
 use crate::models::{
-    QuestDocument, QuestTaskDocument, QuizInsertDocument, QuizQuestionDocument,
+    QuestDocument, QuestInsertDocument, QuestTaskDocument, QuizInsertDocument, QuizQuestionDocument,
```
Comment (Marchand-Nicolas) [inline src/endpoints/admin/quiz/create_question.rs]: No need to import `QuestInsertDocument`

**Review by Marchand-Nicolas — CHANGES_REQUESTED** [review]
A lot of new imports that are not useful, and variables that are now unused

File: src/endpoints/admin/twitter/create_twitter_fw.rs [hunk 10]
```diff
@@ -31,19 +32,16 @@ pub async fn handler(
     // Get the last id in increasing order
     let last_id_filter = doc! {};
     let options = FindOneOptions::builder().sort(doc! {"id": -1}).build();
```
Comment (Marchand-Nicolas) [inline src/endpoints/admin/twitter/create_twitter_fw.rs]: options & last_id_filter unused

File: src/endpoints/admin/quiz/create_question.rs (hunk 9 above)
Comment (Marchand-Nicolas) [reply src/endpoints/admin/quiz/create_question.rs]: Still there

File: src/endpoints/admin/quest_boost/create_boost.rs [hunk 11]
```diff
@@ -44,13 +46,10 @@ pub async fn handler(
     // Get the last id in increasing order
     let last_id_filter = doc! {};
```
Comment (Marchand-Nicolas) [inline src/endpoints/admin/quest_boost/create_boost.rs]: Same options & last_id_filter unused, remove them from all the modified routes

File: src/utils.rs [hunk 12]
```diff
@@ -840,3 +841,28 @@ pub fn parse_string(input: &str, address: FieldElement) -> String {
...
+) -> i32 {
+    let last_id_filter = doc! {};
+    let options = FindOneOptions::builder().sort(doc! {"id": -1}).build();
+
+    let last_doc = task_collection
+        .find_one(last_id_filter, options)
+        .await
+        .unwrap();
+
+    let mut next_id: i32 = 1;
```
Comment (Marchand-Nicolas) [inline src/utils.rs]: Here I have the following warning:

```rs
value assigned to `next_id` is never read
maybe it is overwritten before being read?
`#[warn(unused_assignments)]` on by default
```

I think you can just not define this variable, and replace `next_id = ` by `return`

**Review by Marchand-Nicolas — CHANGES_REQUESTED** [review]

File: src/endpoints/admin/twitter/create_twitter_rw.rs [hunk 13]
```diff
@@ -1,7 +1,8 @@
+use crate::middleware::auth::auth_middleware;
```
Comment (Marchand-Nicolas) [inline src/endpoints/admin/twitter/create_twitter_rw.rs:1]: You can remove `use mongodb::options::FindOneOptions;` from the imports

File: src/endpoints/admin/twitter/create_twitter_fw.rs [hunk 14]
```diff
@@ -31,19 +32,16 @@ pub async fn handler(
     // Get the last id in increasing order
     let last_id_filter = doc! {};
```
Comment (Marchand-Nicolas) [inline src/endpoints/admin/twitter/create_twitter_fw.rs]: Unused variables

File: src/endpoints/admin/discord/create_discord.rs [hunk 15]
```diff
@@ -1,9 +1,10 @@
-use crate::models::{QuestDocument,QuestTaskDocument};
```
Comment (Marchand-Nicolas) [inline src/endpoints/admin/discord/create_discord.rs:1]: Your PR introduces a lot of warnings, please check the compilation before pushing: 
image

**Review by Marchand-Nicolas — CHANGES_REQUESTED** [review]

**Review by Marchand-Nicolas — APPROVED** [review]
Lgtm!

**Pull request merged**